packages/crawler/src/evaluate-page.ts

```diff
@@ -0,0 +1,155 @@
+          }
+          return result + slotEl.outerHTML;
+        }, '' as string) || el.textContent;
+
+      const getHostPdsComponent = (el: Element, prefix: string): TagName | null => {
+        const rootNode = (el.getRootNode() as ShadowRoot).host;
+        if (rootNode) {
+          return getPdsComponentNameByPrefix(rootNode, prefix);
+        }
+        return null;
+      };
+
+      const parseIfJSON = (jsonStr: string | number | symbol): object | string | number | symbol => {
+        if (typeof jsonStr === 'string' && jsonStr.includes('{')) {
+          try {
+            // jsonStr is potentially JSON parsable string, e.g. "{ base: 'block', l: 'inline' }"
+            return JSON.parse(
+              jsonStr
+                .replace(/'/g, '"') // convert single quotes to double quotes
+                .replace(/[\s"]?([a-z\-]+)[\s"]?:([^//])/g, '"$1":$2') // wrap keys in double quotes if they don't have them but ignore potential urls
+            );
+          } catch {
+            // jsonStr is string, e.g. "block" or "inline"
+            return jsonStr;
+          }
+        } else {
+          // jsonStr is object, e.g. { base: 'block', l: 'inline' } or number, e.g. 123 or boolean, e.g. true
+          return jsonStr;
+        }
+      };
+
+      const getAllConsumedProperties = <
+        PComponentName extends keyof HTMLElementTagNameMap,
+        PComponentElement extends HTMLElementTagNameMap[PComponentName],
+        PComponentPropertyName extends keyof PComponentElement,
+        PComponentPropertyValue extends keyof PComponentElement[PComponentPropertyName]
+      >(
+        el: Element,
+        allPdsPropertiesForComponentName: string[]
+      ) => {
+        const pEl = el as PComponentElement;
+
+        const stringifyIfObject = (val: PComponentPropertyValue): PComponentPropertyValue => {
+          const parsedVal = parseIfJSON(val);
+          // check if it's an object and stringify
+          return typeof parsedVal === 'object'
+            ? (JSON.stringify(parsedVal) as PComponentPropertyValue)
+            : (parsedVal as PComponentPropertyValue);
+        };
+
+        return allPdsPropertiesForComponentName.reduce((result, propName) => {
+          const propValue = pEl[propName as PComponentPropertyName] as PComponentPropertyValue;
+          return {
+            ...result,
+            [propName]: stringifyIfObject(propValue),
+          };
+        }, {});
+      };
+
+      const getConsumedTagNamesForPrefix = (prefix: string, pdsElements: Element[]): TagNameData[] => {
+        return pdsElements.map((el) => {
+          const componentName = getPdsComponentNameByPrefix(el, prefix);
+
+          if (!componentName) {
+            throw new Error('Could not find component name');
+          }
+
+          const slotInfo = getChildren(el);
+          const hostPdsComponent = getHostPdsComponent(el, prefix);
+
+          return {
+            [componentName]: {
+              ...{
```

Comment:
Unnecessary object composition.

Reply:
@denyo In a case there's no "slotInfo" or no "hostPdsComponent", I want the returned object not to have the keys "children" and "hostPdsComponent", therefore I use object composition here..

Or is it possible to do it without object composition?

Reply:
I changed "children" and "hostPdsComponent" to be null by default, so could get rid of composition here.